pc/n64dasm/trunk/n64dasm.1
.TH "n64dasm" "1" "1.4" "Marshall B. Rogers" "N64 development tools"
.SH "NAME"
.LP 
\fBn64dasm\fR \-\- Disassemble MIPS R4300i machine code
.SH "SYNOPSIS"
.LP 
n64dasm [\-GuUaHdcLh] [\-E \fIoffset\fP] [\-f \fIfile\fP] [\-O \fIorder\fP] [\-p \fIaddress\fP] [\-S \fIoffset\fP] [\-s \fIspacing\fP] [\-Y \fIsymtab\fP]
.SH "DESCRIPTION"
.LP 
n64dasm is a tiny and flexible disassembler for raw binary blobs containing machine code from the MIPS R4300i processor. It has a myriad of options controlling output, and has primitive symbol table support.
.LP 
Reads/writes from a file/stdin to a file, stdout, or pager.
.SH "OPTIONS"
.LP 
.TP 
\fB\-a\fR
Display the current PC on the left side of the disassembly.

.TP 
\fB\-c\fR
Produce colors in the output disassembly (primative syntax hilighting). If the \fB\-a\fR option is specified, the address will be bold.

.TP 
\fB\-d\fR
Prefix the registers with a dollar sign.

.TP 
\fB\-E\fR \fIoffset\fP
Reads up until \fIoffset\fP in the input file before starting read. This option is invalid if n64dasm is reading from stdin.

.TP 
\fB\-f\fR \fIfile\fP
Uses the specified \fIfile\fP for input. You can also use "\-" to signify standard input.

.TP 
\fB\-G\fR
Parse the input as a GameShark code.

.TP 
\fB\-H\fR
Display the the hexadecimal representation of each instruction on the left side. If \fB\-a\fR is specified, the raw instruction will be printed to the right of the address.

.TP 
\fB\-L\fR
Reroute all output to the pager specified by the environment variable PAGER. If there is none, it will get passed to `less`.

.TP 
\fB\-O\fR \fIorder\fP
Swaps the input from the specified byte \fIorder\fP. Valid orders are: 0123 (default), 3210 (32\-bit little endian), 1032 (16\-bit little endian, or `V64 endian`)

.TP 
\fB\-o\fR \fIfile\fP
Uses the specified \fIfile\fP for output. You can also use "\-" to signify standard input. Using this option in conjunction with \fB\-L\fP is an error.

.TP 
\fB\-p\fR \fIaddress\fP
Sets the PC start to hexadecimal number specified by \fIaddress\fP. The default is 0, and this value is typically set to something like 0x80000000 (the start of the N64's cached RAM).

.TP 
\fB\-S\fR \fIoffset\fP
Seeks to \fIoffset\fP in the input file before starting read. This option is invalid if n64dasm is reading from stdin.

.TP 
\fB\-s\fR \fIspacing\fP
Sets the spacing between opcodes and their operands to \fIspacing\fP.

.TP 
\fB\-U\fR
Print the operands in upper case.

.TP 
\fB\-u\fR
Print the opcodes in upper case.

.TP 
\fB\-Y\fR \fIfile\fP
Uses \fIfile\fR as the symbol table.
.SH "ENVIRONMENT VARIABLES"
.LP 
.TP 
\fBPAGER\fP
When invoked with the \fB\-L\fR flag, \fBn64dasm\fR will pipe the output into the program specified by PAGER. The program must be less or something less\-like, as less specific options are specified by the program (\-SR# 8). If PAGER is unset, \fBless\fR is used.
.SH "EXAMPLES"
.LP 
By default, the program reads from stdin, so if you want to disassemble a file you have sitting around:
.br 
$ n64dasm < mybinary.bin
.LP 
Alternativly you can run it as:
.br 
$ n64dasm \-fL mybinary.bin
.br 
Which will read from the file 'mybinary.bin' and write the output into the pager specified by PAGER.
.LP 
$ n64dasm \-f ROM.bin \-S 0x1000 \-E 0x2000 \-p 0x80000400 \-a
.br 
Will disassemble some code at the entry point.
.LP 
$ n64dasm \-ac \-f ROM.bin \-Y syms.txt \-S 0x1000 \-E 0x2000 \-p 0x80173c60
.br 
Will read from the file ROM.bin, set the PC to 0x80173C60, read symbols from syms.txt, enable the use of colors, enable address output, and adjust the output accordingly.
.SH "AUTHORS"
.LP 
Marshall B. Rogers (mbr@64.vg)
.br 
http://z64dev.sourceforge.net/
.br 
http://www.64.vg/


pc/n64dasm/tags/v1.4/n64dasm.c
/*
*   Nintendo 64 Disassembler -- n64dasm.c
*   Copyright (c) 2009  ZZT32 [mbr@64.vg]
*
*   This program is free software; you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation; either version 2 of the License, or
*   any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public
*   Licence along with this program; if not, write to the Free
*   Software Foundation, Inc., 51 Franklin Street, Fifth Floor, 
*   Boston, MA  02110-1301, USA
*/

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdarg.h>
#include <sys/wait.h>
#include <stdint.h>
#include "decoder.h"
#include "sym.h"
#include "swap.h"
#include "ansi.h"


/* ----------------------------------------------
   Local variables
   ---------------------------------------------- */

static FILE * in;
static FILE * out;
static int ch;
static int fflag, oflag, Yflag, Gflag, uflag, Uflag, sflag, pflag, aflag, Hflag;
static int Sflag, Eflag, dflag, cflag, Lflag, Oflag;
static unsigned w, pc = 0, S, E;
static char * pager = "less";
static uint16_t order;


/* ----------------------------------------------
   Local functions
   ---------------------------------------------- */

static int
STATUSM ( char * fmt, ... )
{
    va_list ap;
    char format[512];
    int len;
    
    sprintf( format, "[" ANSI_SET_FG_BLUE "-" ANSI_RESET_DEFAULT "] %s\n", fmt );
    
    va_start( ap, fmt );
    len = vfprintf( stderr, format, ap );
    va_end( ap );
    
    return len;
}


static void 
show_usage ( char * n )
{
    fprintf
    ( 
       stderr, 
      "usage: %s [-f file] [-o file] [-S offset] [-E offset] [-Y symtab] "
      "[-p address] [-s spacing] [-O order] [-GuUaHdcLh]\n",
      n
    );
    
    exit( EXIT_FAILURE );
}

static void
line_predicate ( unsigned pc, unsigned w, char * dasm )
{
    if( aflag )
    {
        if( cflag )
            fprintf( out, ANSI_SET_FG_WHITE_BOLD "%08X" ANSI_RESET_DEFAULT ": ", pc );
        else
            fprintf( out, "%08X: ", pc );
    }
    
    if( Hflag )
    {
        if( cflag )
            fprintf( out, "("ANSI_SET_FG_CYAN"%08X"ANSI_RESET_DEFAULT") ", w );
        else
            fprintf( out, "(%08X) ", w );
    }
}

static void
print ( unsigned pc, unsigned w, char * dasm )
{
    const char * n;
    
    line_predicate( pc, w, dasm );
    
    if( Yflag )
    {
        if( (n = sym_lookup(pc)) )
        {
            if( !cflag )
                fprintf( out, "%s:\n", n );
            else
                fprintf( out, ANSI_SET_FG_GREEN "%s" ANSI_RESET_DEFAULT ":\n", 
                    n );
            line_predicate( pc, w, dasm );
        }
    }
    
    fprintf( out, "%s%s\n", (Yflag||aflag||Hflag?"        ":""), dasm );
    fflush( out );
}


/* ----------------------------------------------
   Entry point
   ---------------------------------------------- */

int main ( int argc, char ** argv )
{
    static unsigned char buffer[4], dasm[512];
    int fd[2], status;
    in = stdin;
    out = stdout;
    
    STATUSM
    ( 
      ANSI_SET_FG_WHITE_BOLD
      "n64dasm" 
      ANSI_RESET_DEFAULT 
      " version " VERSION "." 
    );
    
    while( (ch = getopt(argc, argv, "f:o:Y:uUs:p:aHS:E:O:dcGL")) != -1 )
    {
        switch( ch )
        {
            case 'f': 
             fflag = 1;
             
             if( !strcmp(optarg, "-") )
                 break;
             
             if( !(in = fopen(optarg, "rb")) )
             {
                 STATUSM( "Unable to access `%s`.", optarg );
                 exit( 1 );
             }
             
             STATUSM( "Reading from `%s`.", optarg );
            break;
            
            case 'o':  
             if( Lflag )
             {
                 STATUSM( "You can't use -L in conjunction with -o." );
                 exit( EXIT_FAILURE );
                 break;
             }
             
             oflag = 1;
             
             if( !strcmp(optarg, "-") )
                 break;
             
             if( !(out = fopen(optarg, "wb")) )
             {
                 STATUSM( "Unable to access `%s`.", optarg );
                 exit( 1 );
             }
             
             STATUSM( "Writing to `%s`.", optarg );
            break;
            
            case 'Y':
             Yflag = 1;
             
             STATUSM( "Reading symbols from `%s`...", optarg );
             STATUSM( "Read %i symbols.", sym_load(optarg) );
            break;
            
            case 's':
             sflag = 1;
             
             mr4kd_spacing( atoi(optarg) );
             
             STATUSM( "Spacing set to %s.", optarg );
            break;
            
            case 'u':
             uflag = 1;
             
             mr4kd_flag_clear( MR4KD_OLOWER );
             
             STATUSM( "Displaying opcodes in upper case." );
            break;
            
            case 'U':
             Uflag = 1;
             
             mr4kd_flag_clear( MR4KD_RLOWER | MR4KD_HLOWER );
             
             STATUSM( "Displaying operands in upper case." );
            break;
            
            case 'p':
             pflag = 1;
                
             if( sscanf( optarg, "%x", &pc ) == 1 )
                STATUSM( "PC set to 0x%08X.", pc );
            break;
            
            case 'a':
             aflag = 1;
             
             STATUSM( "Will display address at left hand side." );
            break;
            
            case 'H':
             Hflag = 1;
             
             STATUSM( "Will display hex representation on left." );
            break;
            
            case 'S':
             Sflag = 1;
             
             sscanf( optarg, "%x", &S );
             
             STATUSM( "Skipping to 0x%08X.", S );
            break;
            
            case 'G':
             Gflag = 1;
             
             STATUSM( "Treating input as a GameShark code.", S );
            break;
            
            case 'E':
             Eflag = 1;
             
             sscanf( optarg, "%x", &E );
             
             STATUSM( "Reading until 0x%08X.", E );
            break;
            
            case 'd':
             dflag = 1;
             
             mr4kd_flag_set( MR4KD_RPREFIX );
             
             STATUSM( "Using dollar sign prefix." );
            break;
            
            case 'c':
             cflag = 1;
             
             mr4kd_flag_set( MR4KD_COLOR );
             
             STATUSM( "Using colors in output." );
            break;
            
            case 'L':
            {
              char * pgr;
              
              if( oflag )
              {
                  STATUSM( "You can't use -o in conjunction with -L." );
                  exit( EXIT_FAILURE );
                  break;
              }
              
              Lflag = 1;
              
              pgr = getenv( "PAGER" );
              
              if( pgr )
                pager = pgr;
              
              STATUSM( "Using `%s` to page output.", pager );
            }
            break;
            
            case 'O':
             Oflag = 1;
             unsigned tmpw;
             
             if( !sscanf( optarg, "%X", &tmpw ) )
             {
                goto invalid_flag;
             }
             
             order = (uint16_t)tmpw;
             
             if( order != 0x3210 && order != 0x1032 && order != 0x0123 )
                 goto invalid_flag;
             
             STATUSM( "Swapping %04X -> %s.", order, "0123" );
             
             break;
             
         invalid_flag:
             STATUSM( "Invalid flag to -O: %s.", optarg );
             Oflag = 0;
            break;
            
            case '?':
            default:
             show_usage( argv[0] );
            break;
        }
    }
    
    /* Start the pager */
    if( Lflag )
    {
        int v;
        
        pipe( fd );
        
        if( !fork() )
        {
            /* Set up pipes */
            close( fd[1] );
            close( 0 );
            dup( fd[0] );
            
            /* Run the pager */
            v = execlp( pager, pager, "-SR#", "8", NULL );
            
            exit( -1 );
        }
        
        close( 1 );
        dup( fd[1] );
        close( fd[0] );
    }
        
    
    if( Sflag )
        fseek( in, S, SEEK_SET );
    
    if( Gflag )
        goto gs;
    
    while( !feof(in) )
    {
        if( Eflag && ftell(in) >= E )
            break;
        
        fread( buffer, 1, 4, in );
        w = buffer[0] << 24 | buffer[1] << 16 | buffer[2] << 8 | buffer[3];
        
        /* Swap the buffer? */
        if( Oflag )
        {
            w = swap_word( w, order );
        }
        
        mr4kd_disassemble( w, pc, (char*)dasm );
        
        print( pc, w, (char*)dasm );
        
        pc += 4;
    }

ret:
    
    if( Lflag )
    {
        fflush( out );
        fclose( out );
        close( fd[1] );
        
        wait( &status );
    }
    
    return 0;
    
gs:
    
    while( !feof(in) )
    {
        unsigned a, v;
        
        if( fscanf( in, "%08X %04X", &a, &v ) != 2 )
            break;
        a &= 0xF0FFFFFF;
        pc = a;
        w = (v << 16);
        if( fscanf( in, "%08X %04X", &a, &v ) != 2 )
            break;
        a &= 0xF0FFFFFF;
        w |= v;
        if( a - pc != 2 )
            STATUSM( "PC jumped from %08X to %08X!", pc, a );
        
        mr4kd_disassemble( w, pc, (char*)dasm );
        
        print( pc, w, (char*)dasm );
    }
    
    goto ret;
}
